File: Python/I/11-ZOO/4/dragdroplineedit.py
from PyQt5.QtCore import *
from PyQt5.QtGui import *
from PyQt5.QtWidgets import *

class DragDropLineEdit(QLineEdit):
    tag = -1
    
    def dragEnterEvent(self, event):
        event.accept()

    def dropEvent(self, event):
        src = event.source()
        if src.tag >= self.tag:
            self.setText(src.text())
            self.tag = src.tag
        src.setVisible(False)

File: C++/I/11-ZOO/4/dragdroplineedit.cpp
#include "dragdroplineedit.h"
#include "dragdroplabel.h"

DragDropLineEdit::DragDropLineEdit(QWidget *parent):
    QLineEdit(parent){}

void DragDropLineEdit::dragEnterEvent(QDragEnterEvent *event)
{
    event->accept();
}

void DragDropLineEdit::dropEvent(QDropEvent *event)
{
    DragDropLabel *src = (DragDropLabel*)event->source();
    if (src->tag() >= tag())
    {
        setText(src->text());
        setTag(src->tag());
    }
    src->setVisible(false);
}
